File: cpp/src/zstd-dict.cc
#include "zstd.h"
#include "zstd-dict.h"


static void CloseCDict(ZSTD_CDict_s* cdict)
{
    ZSTD_freeCDict(cdict);
}


static void CloseDDict(ZSTD_DDict_s* ddict)
{
    ZSTD_freeDDict(ddict);
}


//
// ZstdCompressionDict
//
////////////////////////////////////////////////////////////////////////////////

ZstdCompressionDict::ZstdCompressionDict(const Vec<u8>& dict_bytes, int compression_level)
    : Resource(ZSTD_createCDict(&dict_bytes[0], dict_bytes.size(), compression_level), CloseCDict)
{
}


bool ZstdCompressionDict::fail() const
{
    return get() == nullptr;
}


//
// ZstdDecompressionDict
//
////////////////////////////////////////////////////////////////////////////////

ZstdDecompressionDict ::ZstdDecompressionDict(const Vec<u8>& dict_bytes)
    : Resource(ZSTD_createDDict(&dict_bytes[0], dict_bytes.size()), CloseDDict)
{
}


bool ZstdDecompressionDict::fail() const
{
    return get() == nullptr;
}

File: js/lib/zstd-dict.js
const binding = require('./module.js').Binding;


class ZstdCompressionDict {
    constructor(dict_bytes, compression_level) {
        this.binding = binding.createCompressionDict(dict_bytes, compression_level);
    }

    get() {
        return this.binding;
    }

    close() {
        if (this.binding) {
            this.binding.delete();
        }
    }

    delete() {
        this.close();
    }
}


class ZstdDecompressionDict {
    constructor(dict_bytes) {
        this.binding = new binding.createDecompressionDict(dict_bytes);
    }

    get() {
        return this.binding;
    }

    close() {
        if (this.binding) {
            this.binding.delete();
        }
    }

    delete() {
        this.close();
    }
}


exports.ZstdCompressionDict = ZstdCompressionDict;
exports.ZstdDecompressionDict = ZstdDecompressionDict;
